src/components/shared/Activity/index.tsx

```diff
@@ -1,108 +1,366 @@
-import { Member } from "../../../store/reducers/members";
-import { CurrencyValue } from "../Currency";
-import { RouteName } from "../Navigation";
-
 /**
- * Represents the direction of the relationship between actors in an activity.
- * Currently only supports onward or bidirectional, since that makes the flow
- * of the application linear and therefore more comprehensible.
+ * General purpose component that displays all types of activities. Specific
+ * Activities like GiveOperationActivity ultimately output this component.
  */
-export enum ActivityDirection {
-  Forward = "Forward",
-  Bidirectional = "Bidirectional"
-}
+import * as React from "react";
+import { formatRelative } from "date-fns";
+import { View, StyleSheet, TextStyle, ViewStyle, Image } from "react-native";
+import { withNavigation, NavigationInjectedProps } from "react-navigation";
+import Icon from "react-native-vector-icons/FontAwesome5";
 
-/**
- * Reference to find and render a video.
- */
-interface VideoReference {
-  thumbnailUrl: string;
-  videoUrl: string;
-}
+import { Text } from "../elements";
+import { colors, palette } from "../../../helpers/colors";
+import {
+  VideoWithPlaceholderView,
+  VideoWithPlaceholder
+} from "../VideoWithPlaceholder";
+import {
+  Activity as ActivityData,
+  ActivityContent as ActivityContentData,
+  ActivityCallToAction as CallToActionData,
+  ActivityDirection,
+  BodyType,
+  MediaBody
+} from "../../../store/selectors/activities/types";
+import { MemberName } from "../MemberName";
+import { MemberThumbnail } from "../MemberThumbnail";
+import { Currency } from "../Currency";
+import { TextLink } from "../elements/TextLink";
+import { ArrowHeadDirection, ArrowHead } from "./ArrowHead";
+import { fontSizes } from "../../../helpers/fonts";
 
-/**
- * Reference to find and render an image.
- */
-interface ImageReference {
-  imageUrl: string;
-}
+type Props = {
+  activity: ActivityData;
+};
 
-/**
- * Reference to visual media to dipslay.
- */
-export type MediaReference = VideoReference | ImageReference;
+const CallToAction: React.StatelessComponent<{
+  callToAction: CallToActionData;
+}> = ({ callToAction: { actor, actions } }) => {
+  return (
+    <View>
+      <MemberThumbnail style={styles.actorThumbnail} member={actor} />
+      {actions.map(action => (
+        <View>
+          {action.text.map(piece => {
+            if (typeof piece === "string") {
+              return <Text>piece</Text>;
+            }
+            if ("currencyType" in piece) {
```

Comment:
I feel a bit weird about this mechanism of detecting what type something is and thus how it should be rendered. Maybe each of these action components should have an enum "type" associated with it that tells us how to render?

Reply:
Good point, i'll save that for a future PR

Reply:
and leave a TODO